include/misc/mempool_base.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SYS_MEMPOOL_BASE_H
+#define SYS_MEMPOOL_BASE_H
+
+#include <zephyr/types.h>
+#include <stddef.h>
+
+/*
+ * Definitions and macros used by both the IRQ-safe k_mem_pool and user-mode
+ * compatible sys_mem_pool implementations
+ */
+
+struct sys_mem_pool_lvl {
+	union {
+		u32_t *bits_p;
+		u32_t bits;
+	};
+	sys_dlist_t free_list;
+};
+
+#define SYS_MEM_POOL_KERNEL	0
+#define SYS_MEM_POOL_USER	1
+
+struct sys_mem_pool_base {
+	void *buf;
+	size_t max_sz;
+	u16_t n_max;
+	u8_t n_levels;
+	u8_t max_inline_level;
+	struct sys_mem_pool_lvl *levels;
+	u8_t flags;
```

Comment:
Adding flags there bumps the struct size by a whole word.  If we care (I mean, it's not a lot of memory, but then this is a heap after all) you could steal a bit from n_levels, which by constructions can't be higher than 16 right now (and no higher than 32 even in a fully-addressible 64 bit address space).

Though that has performance implications of course.  Stealing a bit from n_max would be faster, as a quick check shows that that is read only once per free() call.  And except for weird situations where you want really tiny "max size" blocks, 32k is still plenty.

Reply:
I'm confused why this wasn't a packed struct to begin with.
We have in sequence, a u16_t and then two u8_t members which are currently using 32-bits each on probably every architecture.

Reply:
Also worth noting: this is the sys_mem_pool_base struct. There are not going to be too many instances of these in any use-case I can conceive, maybe one for the kernel heap and one for each logical application in the system, maybe one or two extras. So I'd be more inclined to favor performance.